### PR TITLE
Fix reflectometry loader

### LINKED ISSUE #21907: LoadILLReflectometry cannot load future Figaro, D17 NeXus files
Starting from the first cycle in 2018, D17 and Figaro reflectometers will get updated NeXus format. The new format includes:
- Fixed chopper speeds and phases (D17)
- Fixed slit positions (Figaro)
- D0 and D1 fields in Distance block, signifying the source-sample and sample-detector distances

Figaro:
- Find instrument/Distance/edelay_delay not anymore int Theta/; D17 not concerned

### PR DESCRIPTION
**Pull request by vreimund**
Current Figaro data files cannot be loaded since the Nexus format changed.

These changes re-enable loading of all Nexus ILL reflectometry data (which are valid)

In particular the test has many unused lines since the next change of the Nexus format is planned in beginning of September. They were needed for Nexus validation which failed.
D0 and D1 are not calculated correctly for our use. This will be changed in near future.

Reason for intermediate Nexus files seem to be communication problems. I will iteratively verify the new Nexus files beginning of July until September.

**To test:**

Loading of ILL reflectometry test files, please note that the Figaro IDF has changed in terms of its instrument name.

Fixes #21907.

Does this update require release notes?
- [x] Yes

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
@reimundILL There seems to be a conflict now with the sign theta changes in the Figaro definition

**Comment by reimundILL** [comment]
@martyngigg Thanks for informing me

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 1]
```diff
@@ -38,19 +38,19 @@ struct PeakInfo {
  *  @param x an angle in degrees
  *  @return the angle in radians
  */
-constexpr double inRad(const double x) { return x * M_PI / 180; }
+constexpr double inRad(const double x) { return x * M_PI / 180.; }
```
Comment (ianbush) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: I would suggest changing these names to degToRad or similar, else it does not really explain what they are doing

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 2]
```diff
@@ -38,19 +38,19 @@ struct PeakInfo {
...
  */
-constexpr double inDeg(const double x) { return x * 180 / M_PI; }
+constexpr double inDeg(const double x) { return x * 180. / M_PI; }
 
 /** Convert millimeters to meters.
  *  @param x a distance in millimeters
  *  @return the distance in meters
  */
-constexpr double inMeter(const double x) { return x * 1e-3; }
+constexpr double inMeter(const double x) { return x * 1.e-3; }
```
Comment (ianbush) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: As comment above

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 3]
```diff
@@ -38,19 +38,19 @@ struct PeakInfo {
...
  */
-constexpr double inRad(const double x) { return x * M_PI / 180; }
+constexpr double inRad(const double x) { return x * M_PI / 180.; }
 
 /** Convert radians to degrees.
  *  @param x an angle in radians
  *  @return the angle in degrees
  */
-constexpr double inDeg(const double x) { return x * 180 / M_PI; }
+constexpr double inDeg(const double x) { return x * 180. / M_PI; }
```
Comment (ianbush) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: As comment above

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 4]
```diff
@@ -793,13 +782,17 @@ void LoadILLReflectometry::initPixelWidth() {
 /// Update detector position according to data file
 void LoadILLReflectometry::placeDetector() {
   g_log.debug("Move the detector bank \n");
+  // if (m_localWorkspace->run().hasProperty("Distance.D1")) // Valid from 2018.
+  //  m_detectorDistance = inMeter(doubleFromRun("Distance.D1"));
+  // else // Valid before 2018.
+  //  m_detectorDistance = sampleDetectorDistance();
```
Comment (ianbush) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: Is this meant to be commented out? It should be removed or re-instated.

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 5]
```diff
@@ -838,7 +835,12 @@ void LoadILLReflectometry::placeSlits() {
 
 /// Update source position.
 void LoadILLReflectometry::placeSource() {
-  const double dist = sourceSampleDistance();
+  double dist;
+  // if (m_localWorkspace->run().hasProperty("Distance.D0")) // Valid from 2018.
+  //  dist = inMeter(doubleFromRun("Distance.D0"));
+  // else // Valid before 2018.
+  //  dist = sourceSampleDistance();
```
Comment (ianbush) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: Same question on commenting.

File: Framework/DataHandling/test/LoadILLReflectometryTest.h [hunk 6]
```diff
@@ -464,31 +523,219 @@ class LoadILLReflectometryTest : public CxxTest::TestSuite {
...
+    // if (run.hasProperty("Distance.dist_chop_samp"))
+    //  pairCentre = loader.doubleFromRun("Distance.dist_chop_samp");
+    // else
+    //  pairCentre = loader.doubleFromRun("VirtualChopper.dist_chop_samp");
+    // const auto pairSeparation =
+    //    loader.doubleFromRun("Distance.ChopperGap") / 100; // still in cm?
+    // const auto d0ref = pairCentre - 0.5 * pairSeparation;
+    // TS_ASSERT_EQUALS(d0, d0ref);
+    // const auto d0refloader = loader.sourceSampleDistance();
+    // TS_ASSERT_EQUALS(d0, d0refloader);
```
Comment (ianbush) [inline Framework/DataHandling/test/LoadILLReflectometryTest.h]: Also commented out?

File: Framework/DataHandling/test/LoadILLReflectometryTest.h [hunk 7]
```diff
@@ -86,7 +87,7 @@ class LoadILLReflectometryTest : public CxxTest::TestSuite {
     for (const auto &p : properties) {
       loader.setPropertyValue(p.first, p.second);
     }
-    TS_ASSERT_THROWS_NOTHING(loader.execute(););
+    TS_ASSERT_THROWS_NOTHING(loader.execute());
```
Comment (ianbush) [inline Framework/DataHandling/test/LoadILLReflectometryTest.h]: Some semicolons `;` have appeared in after the asserts. These should not be required as `TS_ASSERT_THROWS_NOTHING` are macros.

**Review by ianbush — COMMENTED** [review]
This looks good, I've made some comments that could help clean up the code. I'll try some functional testing next.

I think that this does need adding to the release notes though, as the loader now supports more files.

File: Framework/DataHandling/src/LoadILLReflectometry.cpp (hunk 4 above)
Comment (reimundILL) [reply Framework/DataHandling/src/LoadILLReflectometry.cpp]: Next line calls the `sampleDetectorDistance`. In the future, the Nexus file should contain a correct `D1` entry instead. Then, the method can be deleted

**Review by ianbush — APPROVED** [review]
Looks good to go, merge this once the doc tests have passed :+1: .

**Pull request merged**